### PR TITLE
store/tikv/latch: fix a deadlock in the latch scheduler

### PR DESCRIPTION
**Pull request by tiancaiamao**
### What problem does this PR solve? 

Fix a deadlock in the latch scheduler.
If a transaction release it's lock, it should awake other transactions.
But there is a bug that the transaction does not do that, so some transactions will never get the signal and wait forever.
The blocked transactions will occupy memory and may lead to OOM eventually.

### What is changed and how it works?

When a key release its lock, let the next lock take the key automatically.

### Check List 

Tests 

 - Unit test

Should cherry pick to release-2.1

@disksing @zhangjinpeng1987

### CODE REVIEW

File: store/tikv/latch/scheduler_test.go [hunk 1]
```diff
@@ -28,28 +31,61 @@ func (s *testSchedulerSuite) SetUpTest(c *C) {
...
 }
+
+// generate generates something like:
+// {[]byte("a"), []byte("b"), []byte("c")}
+// {[]byte("a"), []byte("d"), []byte("e"), []byte("f")}
+// {[]byte("e"), []byte("f"), []byte("g"), []byte("h")}
+// The data should not repeat in the sequence.
+func generate() [][]byte {
+	table := []byte{'a', 'b', 'c', 'd', 'e', 'f', 'g', 'h'}
+	xxx := make([]byte, 0, 5)
```
Comment (disksing) [inline store/tikv/latch/scheduler_test.go]: Maybe another name instead of `xxx`?

File: store/tikv/latch/scheduler_test.go [hunk 2]
```diff
@@ -28,28 +31,61 @@ func (s *testSchedulerSuite) SetUpTest(c *C) {
...
+	chance := []int{100, 60, 40, 20}
+	for i := 0; i < len(chance); i++ {
+		needMore := rand.Intn(100) < chance[i]
+		if needMore {
+			randChar := table[rand.Intn(len(table))]
+			if !bytes.Contains(xxx, []byte{randChar}) {
```
Comment (disksing) [inline store/tikv/latch/scheduler_test.go]: `if bytes.IndexByte(xxx, randChar) == -1`

File: store/tikv/latch/scheduler_test.go [hunk 3]
```diff
@@ -28,28 +31,61 @@ func (s *testSchedulerSuite) SetUpTest(c *C) {
...
+				xxx = append(xxx, randChar)
+			}
+		}
+	}
+
+	// Turn []byte{"x","y","z"} into [][]byte{[]byte("x"), []byte("y"), []byte("z")}
```
Comment (disksing) [inline store/tikv/latch/scheduler_test.go]: Why not create `xxx` as `[][]byte` in at beginning?

**Comment by tiancaiamao** [comment]
1. Let the staled Lock obtain the slot in `releaseSlot` to fix the deadlock
2. Remove the transaction memory's reference from `node`
3. Adjust the frequency setting so the memory can be recycled timely

PTAL @zhangjinpeng1987 @disksing

**Comment by disksing** [comment]
LGTM.

**Comment by zhangjinpeng1987** [comment]
I think you can add a long time running test to Schrodinger, to discover potential memory leak issues.

**Comment by zhangjinpeng1987** [comment]
LGTM

**Pull request merged**